src/semanticlayertools/linkage/cocitation.py

```diff
@@ -0,0 +1,171 @@
+    :type columnName: str
+    :param numberProc: Number of CPUs the package is allowed to use (default=all)
+    :type numberProc: int
+    :param limitRefLength: Either False or integer giving the maximum number of references a considered publication is allowed to contain
+    :type limitRefLength: bool or int
+    :param timerange: Time range to consider (default=(1945,2005))
+    :type timerange: tuple
+    :param debug: False/True or l2 to show level 2 debugging messages
+    """
+
+    def __init__(
+        self, inpath, outpath, columnName,
+        numberProc: int = num_processes,
+        limitRefLength: limitRefLength = False,
+        timerange: tuple = (1945, 2005),
+        debug: debugVar = False
+    ):
+        self.inpath = inpath
+        self.outpath = outpath
+        self.columnName = columnName
+        self.numberProc = numberProc
+        self.limitRefLength = limitRefLength
+        self.timerange = timerange
+        self.debug = debug
+
+    def getCombinations(self, chunk):
+        """Calculate combinations of references in publications chunk.
+
+        :param chunk: A chunk of the corpus dataframe
+        :type chunk: `pd.Dataframe`
+        :returns: A list of all reference combinations for each corpus entry
+        :rtype: list
+        """
+        res = []
+        if type(self.limitRefLength) == int:
+            reflen = chunk[self.columnName].apply(
+                lambda x: True if type(x) == list and len(x) <= self.limitRefLength else False
+            )
+            data = chunk[reflen].copy()
+        else:
+            data = chunk.copy()
+        for idx, row in data.iterrows():
+            comb = combinations(row[self.columnName], 2)
+            for elem in list(comb):
+                res.append((elem))
+        return res
+
+    def calculateCoCitation(self, filepath):
+        """Run calculation for single input file.
+
+        Creates three files: Metadata-File with all components information,
+        Giant component network data in pajek format and full graph data in
+        edgelist format.
+
+        :param filepath: Path for input corous
+        :type filepath: str
+        :returns: A tuple of GC information: Number of nodes and percentage of total, Number of edges and percentage of total
+        :rtype: tuple
+        """
+        infilename = filepath.split(os.path.sep)[-1].split('.')[0]
+        starttime = time.time()
+        try:
+            data = pd.read_json(filepath, lines=True).dropna(subset=[self.columnName])
+            chunk_size = int(data.shape[0] / self.numberProc)
+            if chunk_size == 0:  # Deal with small data samples.
+                chunk_size = 1
+            chunks = np.array_split(data, chunk_size)
+            pool = multiprocessing.Pool(processes=self.numberProc)
+            cocitations = pool.map(self.getCombinations, chunks)
+            cocitCounts = Counter([x for y in cocitations for x in y])
```

Comment:
Most of the code from here down is very unclear, comments would have helped me figure out what's going on.

Reply:
you mean inline comments? will add them